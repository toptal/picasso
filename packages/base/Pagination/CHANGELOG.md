# @toptal/picasso-pagination

## 1.0.2

### Patch Changes

- Updated dependencies []:
  - @toptal/picasso-button@1.0.2

## 1.0.1

### Patch Changes

- [#4164](https://github.com/toptal/picasso/pull/4164) [`84d579cee65c88aeb6ba58318c43d8deb5201b30`](https://github.com/toptal/picasso/commit/84d579cee65c88aeb6ba58318c43d8deb5201b30) Thanks [@mkrl](https://github.com/mkrl)!
- migrate to Picasso by Parts
  - picasso is now distributed as a set of independent packages with the main package `@toptal/picasso` now being a collection of re-exported packages
- Updated dependencies [[`84d579cee65c88aeb6ba58318c43d8deb5201b30`](https://github.com/toptal/picasso/commit/84d579cee65c88aeb6ba58318c43d8deb5201b30)]:
  - @toptal/picasso-typography@1.0.1
  - @toptal/picasso-container@1.0.1
  - @toptal/picasso-button@1.0.1
  - @toptal/picasso-utils@1.0.1
