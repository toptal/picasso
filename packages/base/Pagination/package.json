{
  "name": "@toptal/picasso-pagination",
  "version": "4.0.2",
  "description": "Toptal UI components library - Pagination",
  "publishConfig": {
    "access": "public"
  },
  "main": "./dist-package/src/index.js",
  "module": "./dist-package/src/index.js",
  "scripts": {
    "build:package": "tsc -b tsconfig.json",
    "prepublishOnly": "yarn build:package"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/toptal/picasso.git"
  },
  "author": "Toptal",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/toptal/picasso/issues"
  },
  "homepage": "https://github.com/toptal/picasso/tree/master/packages/picasso#readme",
  "dependencies": {
    "@toptal/picasso-button": "4.0.2",
    "@toptal/picasso-container": "3.0.1",
    "@toptal/picasso-typography": "4.0.1",
    "@toptal/picasso-utils": "2.0.0"
  },
  "sideEffects": [
    "**/styles.ts",
    "**/styles.js"
  ],
  "peerDependencies": {
    "react": ">=16.12.0 < 19.0.0",
    "@toptal/picasso-tailwind-merge": "^2.0.0",
    "@toptal/picasso-tailwind": ">=2.7"
  },
  "exports": {
    ".": "./dist-package/src/index.js"
  },
  "devDependencies": {
    "@toptal/picasso-tailwind-merge": "2.0.1",
    "@toptal/picasso-test-utils": "1.1.1"
  },
  "files": [
    "dist-package/**",
    "!dist-package/tsconfig.tsbuildinfo",
    "src"
  ]
}
