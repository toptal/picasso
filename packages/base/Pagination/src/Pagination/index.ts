import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Pagination'

export { default as Pagination } from './Pagination'
export type PaginationProps = OmitInternalProps<Props>
