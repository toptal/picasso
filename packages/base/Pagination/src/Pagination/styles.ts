import { createStyles } from '@material-ui/core/styles'

export default () =>
  createStyles({
    ellipsis: {
      padding: '0 0.5em',
      cursor: 'default',
    },
  })
