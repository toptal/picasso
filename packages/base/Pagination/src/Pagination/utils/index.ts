export { default as getRange, ELLIPSIS } from './get-range'
