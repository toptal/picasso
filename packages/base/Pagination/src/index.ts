export * from './Pagination'
export * from './PaginationButton'