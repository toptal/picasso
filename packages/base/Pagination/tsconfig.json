{
  "extends": "../../../tsconfig.base.json",
  "compilerOptions": { "outDir": "dist-package" },
  "include": ["src"],
  "references": [
    { "path": "../Button" },
    { "path": "../Container" },
    { "path": "../Typography" },
    { "path": "../Utils" }
  ]
}
