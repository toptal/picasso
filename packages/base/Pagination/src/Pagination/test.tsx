import React from 'react'
import { render, fireEvent } from '@toptal/picasso-test-utils'
import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Pagination'
import { Pagination } from './Pagination'

const renderPagination = (props: OmitInternalProps<Props>) => {
  const { activePage, disabled, onPageChange, totalPages } = props

  return render(
    <Pagination
      activePage={activePage}
      disabled={disabled}
      onPageChange={onPageChange}
      totalPages={totalPages}
    />
  )
}

describe('Pagination', () => {
  it('renders', () => {
    const { container } = renderPagination({
      activePage: 5,
      totalPages: 20,
      onPageChange: () => {},
    })

    expect(container).toMatchSnapshot()
  })

  it('renders disabled', () => {
    const { container } = renderPagination({
      activePage: 5,
      totalPages: 20,
      disabled: true,
      onPageChange: () => {},
    })

    expect(container).toMatchSnapshot()
  })

  it('renders nothing for 1 page', () => {
    const { container } = renderPagination({
      activePage: 1,
      totalPages: 1,
      onPageChange: () => {},
    })

    expect(container).toMatchSnapshot()
  })

  it('renders nothing for 0 pages', () => {
    const { container } = renderPagination({
      activePage: 1,
      totalPages: 0,
      onPageChange: () => {},
    })

    expect(container).toMatchSnapshot()
  })

  it('ignores prev click when on first page', () => {
    const onPageChange = jest.fn()
    const { getByText } = renderPagination({
      activePage: 1,
      totalPages: 5,
      onPageChange,
    })

    fireEvent.click(getByText('Prev'))

    expect(onPageChange).not.toHaveBeenCalled()
  })

  it('ignores next click when on last page', () => {
    const onPageChange = jest.fn()
    const { getByText } = renderPagination({
      activePage: 5,
      totalPages: 5,
      onPageChange,
    })

    fireEvent.click(getByText('Next'))

    expect(onPageChange).not.toHaveBeenCalled()
  })
})
