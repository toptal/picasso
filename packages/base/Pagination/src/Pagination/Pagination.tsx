/* eslint-disable react/no-array-index-key */
import type { HTMLAttributes } from 'react'
import React, { forwardRef, useMemo } from 'react'
import type { BaseProps } from '@toptal/picasso-shared'
import { Button } from '@toptal/picasso-button'
import { Container } from '@toptal/picasso-container'
import { Typography } from '@toptal/picasso-typography'

import { getRange, ELLIPSIS } from './utils'
import { PaginationButton } from '../PaginationButton'

export interface Props extends BaseProps, HTMLAttributes<HTMLDivElement> {
  /** Value of the current highlighted page */
  activePage: number
  /** Shows `Pagination` in disabled state when pages are not changeable */
  disabled?: boolean
  /** Callback invoked when any page number is clicked */
  onPageChange: (page: number) => void
  /** Value of total pages of the data set used for calculation of page buttons */
  totalPages: number
  /** Number of the active page siblings  */
  siblingCount?: number
  /** Variant of the pagination representation  */
  variant?: 'default' | 'compact'
}

export const Pagination = forwardRef<HTMLDivElement, Props>(function Pagination(
  props,
  ref
) {
  const {
    activePage,
    disabled,
    totalPages,
    onPageChange,
    siblingCount = 2,
    variant,
    ...rest
  } = props

  const pages = useMemo(
    () => getRange({ activePage, totalPages, siblingCount }),
    [activePage, totalPages, siblingCount]
  )

  if (pages.length <= 1) {
    return null
  }

  const isFirstActive = activePage === 1
  const isLastActive = activePage === totalPages

  const handlePrevClick = () => onPageChange(activePage - 1)
  const handleNextClick = () => onPageChange(activePage + 1)

  const pageButtons = pages.map((page, index) => {
    if (page === ELLIPSIS) {
      return (
        <Container key={index} className='py-0 px-[0.5em] cursor-default'>
          <Typography size='xsmall' weight='semibold' color='black'>
            {ELLIPSIS}
          </Typography>
        </Container>
      )
    }

    return (
      <PaginationButton
        key={index}
        className='[&+&]:ml-2'
        page={Number(page)}
        activePage={activePage}
        disabled={disabled}
        onClick={onPageChange}
      />
    )
  })

  return (
    <Container {...rest} ref={ref} flex inline alignItems='center'>
      <Button
        className='[&+&]:ml-2'
        disabled={isFirstActive || disabled}
        onClick={handlePrevClick}
        variant='secondary'
        size='small'
      >
        Prev
      </Button>

      {variant === 'compact' ? null : pageButtons}

      <Button
        className='[&+&]:ml-2'
        disabled={isLastActive || disabled}
        onClick={handleNextClick}
        variant='secondary'
        size='small'
      >
        Next
      </Button>
    </Container>
  )
})

Pagination.defaultProps = {
  disabled: false,
  siblingCount: 2,
  variant: 'default',
}

Pagination.displayName = 'Pagination'

export default Pagination
