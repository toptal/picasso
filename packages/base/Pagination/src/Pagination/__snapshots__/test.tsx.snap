// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline"
    >
      <button
        class="MuiButtonBase-root PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPagination-button"
        data-component-type="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          Prev
        </span>
      </button>
      <button
        aria-current="false"
        class="MuiButtonBase-root PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPagination-button"
        data-component-type="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          1
        </span>
      </button>
      <div
        class="PicassoPagination-ellipsis"
      >
        <p
          class="m-0 text-xxs text-black font-semibold"
        >
          ...
        </p>
      </div>
      <button
        aria-current="false"
        class="MuiButtonBase-root PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPagination-button"
        data-component-type="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          3
        </span>
      </button>
      <button
        aria-current="false"
        class="MuiButtonBase-root PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPagination-button"
        data-component-type="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          4
        </span>
      </button>
      <button
        aria-current="true"
        class="MuiButtonBase-root PicassoButton-active PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPaginationButton-active PicassoPagination-button"
        data-component-type="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          5
        </span>
      </button>
      <button
        aria-current="false"
        class="MuiButtonBase-root PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPagination-button"
        data-component-type="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          6
        </span>
      </button>
      <button
        aria-current="false"
        class="MuiButtonBase-root PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPagination-button"
        data-component-type="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          7
        </span>
      </button>
      <div
        class="PicassoPagination-ellipsis"
      >
        <p
          class="m-0 text-xxs text-black font-semibold"
        >
          ...
        </p>
      </div>
      <button
        aria-current="false"
        class="MuiButtonBase-root PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPagination-button"
        data-component-type="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          20
        </span>
      </button>
      <button
        class="MuiButtonBase-root PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPagination-button"
        data-component-type="button"
        tabindex="0"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          Next
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Pagination renders disabled 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline"
    >
      <button
        class="MuiButtonBase-root PicassoButton-disabled PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPagination-button Mui-disabled"
        data-component-type="button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          Prev
        </span>
      </button>
      <button
        aria-current="false"
        class="MuiButtonBase-root PicassoButton-disabled PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPagination-button Mui-disabled"
        data-component-type="button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          1
        </span>
      </button>
      <div
        class="PicassoPagination-ellipsis"
      >
        <p
          class="m-0 text-xxs text-black font-semibold"
        >
          ...
        </p>
      </div>
      <button
        aria-current="false"
        class="MuiButtonBase-root PicassoButton-disabled PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPagination-button Mui-disabled"
        data-component-type="button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          3
        </span>
      </button>
      <button
        aria-current="false"
        class="MuiButtonBase-root PicassoButton-disabled PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPagination-button Mui-disabled"
        data-component-type="button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          4
        </span>
      </button>
      <button
        aria-current="true"
        class="MuiButtonBase-root PicassoButton-active PicassoButton-disabled PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPaginationButton-active PicassoPagination-button Mui-disabled"
        data-component-type="button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          5
        </span>
      </button>
      <button
        aria-current="false"
        class="MuiButtonBase-root PicassoButton-disabled PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPagination-button Mui-disabled"
        data-component-type="button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          6
        </span>
      </button>
      <button
        aria-current="false"
        class="MuiButtonBase-root PicassoButton-disabled PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPagination-button Mui-disabled"
        data-component-type="button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          7
        </span>
      </button>
      <div
        class="PicassoPagination-ellipsis"
      >
        <p
          class="m-0 text-xxs text-black font-semibold"
        >
          ...
        </p>
      </div>
      <button
        aria-current="false"
        class="MuiButtonBase-root PicassoButton-disabled PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPaginationButton-root PicassoPagination-button Mui-disabled"
        data-component-type="button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          20
        </span>
      </button>
      <button
        class="MuiButtonBase-root PicassoButton-disabled PicassoButton-small PicassoButton-secondary PicassoButton-root PicassoPagination-button Mui-disabled"
        data-component-type="button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
        >
          Next
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Pagination renders nothing for 0 pages 1`] = `
<div>
  <div
    class="Picasso-root"
  />
</div>
`;

exports[`Pagination renders nothing for 1 page 1`] = `
<div>
  <div
    class="Picasso-root"
  />
</div>
`;
